pym/portage/meta_manifest.py

```diff
@@ -0,0 +1,275 @@
+		if gpgcmd in [None, '']:
+			raise portage.exception.MissingParameter(
+				"PORTAGE_GPG_SIGNING_COMMAND is unset! Is make.globals missing?")
+		if "${PORTAGE_GPG_KEY}" in gpgcmd and "PORTAGE_GPG_KEY" not in portage_settings:
+			raise portage.exception.MissingParameter("PORTAGE_GPG_KEY is unset!")
+		if "${PORTAGE_GPG_DIR}" in gpgcmd:
+			if "PORTAGE_GPG_DIR" not in portage_settings:
+				portage_settings["PORTAGE_GPG_DIR"] = os.path.expanduser("~/.gnupg")
+			else:
+				portage_settings["PORTAGE_GPG_DIR"] = os.path.expanduser(portage_settings["PORTAGE_GPG_DIR"])
+			if not os.access(portage_settings["PORTAGE_GPG_DIR"], os.X_OK):
+				raise portage.exception.InvalidLocation(
+					"Unable to access directory: PORTAGE_GPG_DIR='%s'" %
+					portage_settings["PORTAGE_GPG_DIR"])
+		gpgvars = {"FILE": filename}
+		for setting in ("PORTAGE_GPG_DIR", "PORTAGE_GPG_KEY"):
+			keyvar = portage_settings.get(setting)
+			if keyvar is not None:
+				gpgvars[setting] = keyvar
+		gpgcmd = portage.util.varexpand(gpgcmd, mydict=gpgvars)
+		gpgcmd = portage.util.shlex_split(gpgcmd)
+		gpgcmd = [portage._unicode_encode(arg, encoding=portage._encodings['fs'], errors='strict') for arg in gpgcmd]
+		return_code = subprocess.call(gpgcmd)
+		if return_code == os.EX_OK:
+			os.rename(filename + ".asc", filename)
+			self.signed = True
+		else:
+			raise portage.exception.PortageException("!!! gpg exited with '" + str(return_code) + "' status")
+
+	def find_reporoot(self):
+		'''Finds reporoot'''
+		repodir = self.pkgdir
+		while len(set(['metadata', 'eclass', 'profiles']).intersection(os.listdir(repodir))) == 0:
+			myrepo = repodir.split(os.path.sep)[1:-1]
+			if myrepo[-1] not in myrepo[:-1]:
+				repodir = repodir[:-(len(myrepo[-1])+1)]
+		self.reporoot = repodir
+
+	def find_repolevel(self, path):
+		'''Finds repolevel'''
+		self.find_reporoot()
+		#path = path.replace(self.reporoot, "")
+		path = path[len(self.reporoot):]
+		myrepo = path.split(os.path.sep)
+		repolevel = len(myrepo)
+		return repolevel
+
+	def find_repotype(self, path):
+		'''Finds the type of the repo directory'''
+		if len(set(['metadata', 'eclass', 'profiles']).intersection(os.listdir(path))) > 0:
+			return 'root'
+		elif path.endswith("eclass/"):
+			return 'eclass'
+		elif path.endswith("profiles/"):
+			return 'profile'
+		else:
+			if self.find_repolevel(path) == 2:
+				return 'category'
+			else:
+				return 'package'
+
+	def create(self):
+		'''Creates a MetaManifest file'''
+		repodir = self.pkgdir
+		repotype = self.find_repotype(repodir)
+		if repotype == 'profile':
+			self.create_profile()
+		elif repotype == 'eclass':
+			self.create_eclass()
+		elif repotype == 'root':
+			self.create_master()
+		else:
+			self.create_cat()
+
+	def create_master(self):
+		'''Creates a MetaManifest file from all the MetaManifest files of the repository'''
+		self.find_reporoot()
+		myrepodir = self.reporoot
+		self.pkgdir = self.reporoot
+		self.fhashdict = {}
```

Comment:
You can probably replace this with a `collections.defaultdict(dict)` unless you need to get `KeyError` when accessing an unknown `ftype`.
Also this particular bit is used in multiple `create_` functions, maybe you want to put this in __init__ ? If not, you probably don't want to save this as an instance variable.